packages/compiler/src/core/program.ts

```diff
@@ -151,7 +157,35 @@ export async function compile(
   mainFile: string,
   options: CompilerOptions = {},
   oldProgram?: Program, // NOTE: deliberately separate from options to avoid memory leak by chaining all old programs together.
-): Promise<Program> {
+) {
+  const logger = createLogger({ sink: host.logSink });
+
+  let program, shouldEmit;
+  if (logger.trackAction) {
+    ({ program, shouldEmit } = await logger.trackAction(
+      () => runCompiler(host, mainFile, options, oldProgram),
+      { level: "trace", message: "Compiling..." },
+    ));
+  } else {
+    ({ program, shouldEmit } = await runCompiler(host, mainFile, options, oldProgram));
+  }
+
+  // Emitter stage
+  if (shouldEmit) {
+    for (const emitter of program.emitters) {
+      await emit(emitter, options, program);
+    }
+  }
+
+  return program;
+}
+
+export async function runCompiler(
+  host: CompilerHost,
+  mainFile: string,
+  options: CompilerOptions = {},
+  oldProgram?: Program,
+): Promise<{ program: Program; shouldEmit: boolean }> {
   const validateCbs: Validator[] = [];
   const stateMaps = new Map<symbol, StateMap>();
   const stateSets = new Map<symbol, StateSet>();
```

Comment:
I think we might want to reorganize this function, the `runCompiler` might just want to do any job after the `program` object was created. Just because here it initialize some state that should really be available in every other function of this program and this might be quite confusing.

Reply:
This is a little bit more complex than it looks like, all the functions used by runCompiler  are inside the function and they are using things defined there, I can reorganize all this code but can I do it on a different pr? just to be able to focus on the specific things I am moving as I will have to move a lot of stuff to keep this running